gdx/src/com/badlogic/gdx/graphics/g2d/ArrayTextureSpriteBatch.java

```diff
@@ -0,0 +1,1518 @@
+		totalRenderCalls++;
+
+		int spritesInBatch = idx / (spriteFloatSize + 4);
+		if (spritesInBatch > maxSpritesInBatch) maxSpritesInBatch = spritesInBatch;
+		int count = spritesInBatch * 6;
+
+		if (useMipMaps && mipMapsDirty) {
+			Gdx.gl30.glGenerateMipmap(GL30.GL_TEXTURE_2D_ARRAY);
+			mipMapsDirty = false;
+		}
+
+		Mesh mesh = this.mesh;
+
+		mesh.setVertices(vertices, 0, idx);
+
+		mesh.getIndicesBuffer().position(0);
+		mesh.getIndicesBuffer().limit(count);
+
+		if (blendingDisabled) {
+			Gdx.gl.glDisable(GL30.GL_BLEND);
+		} else {
+			Gdx.gl.glEnable(GL30.GL_BLEND);
+			if (blendSrcFunc != -1) Gdx.gl.glBlendFuncSeparate(blendSrcFunc, blendDstFunc, blendSrcFuncAlpha, blendDstFuncAlpha);
+		}
+
+		if (customShader != null) {
+			mesh.render(customShader, GL30.GL_TRIANGLES, 0, count);
+		} else {
+			mesh.render(shader, GL30.GL_TRIANGLES, 0, count);
+		}
+
+		idx = 0;
+	}
+
+	/** Assigns space on the Array Texture, sets up Texture scaling and manages the LFU cache.
+	 * @param texture The texture that shall be loaded into the cache, if it is not already loaded.
+	 * @return The texture slot that has been allocated to the selected texture */
+	private int activateTexture (Texture texture) {
+
+		subImageScaleWidth = texture.getWidth() * invMaxTextureWidth;
+		subImageScaleHeight = texture.getHeight() * invMaxTextureHeight;
+
+		if (subImageScaleWidth > 1f || subImageScaleHeight > 1f) {
+			throw new IllegalStateException(
+				"Texture " + texture.getTextureObjectHandle() + " is larger than the Array Texture: [" + texture.getWidth() + ","
+					+ texture.getHeight() + "] > [" + (int)(1f / invMaxTextureWidth) + "," + (int)(1f / invMaxTextureHeight) + "]");
+		}
+
+		invTexWidth = subImageScaleWidth / texture.getWidth();
+		invTexHeight = subImageScaleHeight / texture.getHeight();
+
+		// This is our identifier for the textures
+		final int textureHandle = texture.getTextureObjectHandle();
+
+		// First try to see if the texture is already cached
+		for (int i = 0; i < currentTextureLFUSize; i++) {
+
+			// getTextureObjectHandle() just returns an int,
+			// it's fine to call this method instead of caching the value.
+			if (textureHandle == usedTextures[i].getTextureObjectHandle()) {
+
+				// Increase the access counter.
+				usedTexturesLFU[i]++;
+
+				return i;
+			}
+		}
+
+		// If a free texture unit is available we just use it
+		// If not we have to flush and then throw out the least accessed one.
+		if (currentTextureLFUSize < maxTextureLevels) {
+
+			// Put the texture into the next free slot
+			usedTextures[currentTextureLFUSize] = texture;
+
+			// Increase the access counter.
+			usedTexturesLFU[currentTextureLFUSize]++;
+
+			copyTextureIntoArrayTexture(texture, currentTextureLFUSize);
+
+			return currentTextureLFUSize++;
+
+		} else {
+
+			// We have to flush if there is something in the pipeline already,
+			// otherwise the texture index of previously rendered sprites gets invalidated
+			if (idx > 0) {
+				flush();

```

Comment:
This could maybe get optimized a bit by checking which textures are currently used, but it is probably not important.

Reply:
I'll have to rethink the whole cache logic a little since it's changed so much over all the iterations.